google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java

```diff
@@ -0,0 +1,246 @@
+    assume()
+        .withMessage("DirectPath integration tests can only run against DirectPathEnv")
+        .that(testEnvRule.env())
+        .isInstanceOf(DirectPathEnv.class);
+
+    // Create a transport channel provider that can intercept ipv6 packets.
+    channelFactory = new MyChannelFactory();
+    eventLoopGroup = new NioEventLoopGroup();
+
+    BigtableDataSettings defaultSettings = testEnvRule.env().getDataClientSettings();
+    InstantiatingGrpcChannelProvider defaultTransportProvider =
+        (InstantiatingGrpcChannelProvider)
+            defaultSettings.getStubSettings().getTransportChannelProvider();
+    InstantiatingGrpcChannelProvider instrumentedTransportChannelProvider =
+        defaultTransportProvider
+            .toBuilder()
+            .setPoolSize(1)
+            .setChannelConfigurator(
+                new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+                  @Override
+                  public ManagedChannelBuilder apply(ManagedChannelBuilder builder) {
+                    injectNettyChannelHandler(builder);
+
+                    // Fail fast when blackhole is active
+                    builder.keepAliveTime(1, TimeUnit.SECONDS);
+                    builder.keepAliveTimeout(1, TimeUnit.SECONDS);
+                    return builder;
+                  }
+                })
+            .build();
+
+    // Inject the instrumented transport provider into a new client
+    BigtableDataSettings.Builder settingsBuilder =
+        testEnvRule.env().getDataClientSettings().toBuilder();
+
+    settingsBuilder
+        .stubSettings()
+        .setTransportChannelProvider(instrumentedTransportChannelProvider);
+
+    instrumentedClient = BigtableDataClient.create(settingsBuilder.build());
+  }
+
+  @After
+  public void teardown() {
+    if (instrumentedClient != null) {
+      instrumentedClient.close();
+    }
+    if (eventLoopGroup != null) {
+      eventLoopGroup.shutdownGracefully();
+    }
+  }
+
+  @Test
+  public void testFallback() throws InterruptedException {
+    // Verify that we can send a request
+    instrumentedClient.readRow(testEnvRule.env().getTableId(), "nonexistent-row");
+
+    // Enable the blackhole, which will prevent communication via IPv6 and thus DirectPath.
+    blackholeIPv6.set(true);
+
+    // Send a request, which should be routed over IPv4 and CFE.
+    instrumentedClient.readRow(testEnvRule.env().getTableId(), "nonexistent-row");
+
+    // Verify that the above check was meaningful, by verifying that the blackhole actually dropped
+    // packets.
+    assertThat(numBlocked.get()).isGreaterThan(0);
+
+    // Make sure that the client will start reading from IPv6 again by sending new requests and
+    // checking the injected IPv6 counter has been updated.
+    blackholeIPv6.set(false);
+    numIPv6Read.set(0);
+
+    // Arbitrary count that will ensure that there have been enough activity via IPv6.
+    final int MIN_COMPLETE_READ_CALLS = 40;
```

Comment:
Local variables shouldn't be final unless required and shouldn't be named like fields

Reply:
I promoted it to a static field